### PR TITLE
[Core] Socket creation race condition bug fixes

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This fixes 2 issues related to creating sockets.

1. Don't log errors due to a race condition in raylet creating a socket + worker trying to connect.
2. When a worker tries to prepare a socket's directory it should be idempotent.

## Related issue number

#9761

#9763

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by wuisawesome** [comment]
@mehrdadn can you just make sure the `_prepare_socket_file` behavior for windows is still correct?

File: src/ray/common/client_connection.cc [hunk 1]
```diff
@@ -47,7 +47,13 @@ Status ConnectSocketRetry(local_stream_socket &socket, const std::string &endpoi
     if (!ec) {
       break;
     }
-    if (num_attempts > 0) {
+    if (0 < num_attempts && num_attempts < num_retries / 2) {
```
Comment (ijrsvt) [inline src/ray/common/client_connection.cc:50]: nit: could you keep it as `num_attempts > 0`.
Thanks for addressing this :)

**Review by rkooo567 — APPROVED** [review]

**Review by mehrdadn — COMMENTED** [review]
@wuisawesome I don't believe you've touched the Windows path at all (`if sys.platform == "win32":`), so it seems correct to me?

**Comment by wuisawesome** [comment]
> @wuisawesome I don't believe you've touched the Windows path at all (`if sys.platform == "win32":`), so it seems correct to me?

I didn't touch the windows path but I redefined the behavior of the function slightly.

File: src/ray/common/client_connection.cc [hunk 2]
```diff
@@ -47,7 +47,13 @@ Status ConnectSocketRetry(local_stream_socket &socket, const std::string &endpoi
...
+      // Socket is created by the raylet. Due to a race condition it might not
+      // be created before we try connecting.
+      RAY_LOG(INFO) << "Retrying to connect to socket for endpoint " << endpoint
+                    << " (num_attempts = " << num_attempts
+                    << ", num_retries = " << num_retries << ")";
+    } else if (num_attempts >= num_retries / 2) {
       RAY_LOG(ERROR) << "Retrying to connect to socket for endpoint " << endpoint
```
Comment (ericl) [inline src/ray/common/client_connection.cc:57]: Can we just change this to INFO instead of adding that logic above?

File: python/ray/node.py [hunk 3]
```diff
@@ -472,7 +472,8 @@ def _prepare_socket_file(self, socket_path, default_prefix):
 
         This method helps to prepare a socket file.
         1. Make the directory if the directory does not exist.
-        2. If the socket file exists, raise exception.
+        2. If the socket file exists, do nothing (this just means we aren't the
```
Comment (ericl) [inline python/ray/node.py:475]: Why these changes? Isn't suppressing the log message enough?

File: python/ray/tests/test_advanced_3.py [hunk 4]
```diff
@@ -421,7 +421,14 @@ def test_socket_dir_not_existing(shutdown_only):
                                               "tests", random_name)
         temp_raylet_socket_name = os.path.join(temp_raylet_socket_dir,
                                                "raylet_socket")
-        ray.init(num_cpus=1, raylet_socket_name=temp_raylet_socket_name)
+        ray.init(num_cpus=2, raylet_socket_name=temp_raylet_socket_name)
```
Comment (ericl) [inline python/ray/tests/test_advanced_3.py:424]: Why do we need this change?

**Review by ericl — CHANGES_REQUESTED** [review]

File: src/ray/common/client_connection.cc (hunk 1 above)
Comment (wuisawesome) [reply src/ray/common/client_connection.cc]: At risk of starting a flame war, this is soo much closer to `0 < num_attempts < num_retries / 2`. Nevertheless, I've changed it to just never error log :p

**Review by ericl — APPROVED** [review]
